# Iteration - Looping Fundamentals

*PY4E Textbook*


## Updating variables
A common pattern in assignment statements is an assignment statement that updates a variable, where the new value of the variable depends on the old.

In [ ]:
x=10
x=x+1
print(x)


11


This means “get the current value of `x`, add 1, and then update `x` with the new value.”

To make the value of `x` continuously increase, you can use a **loop**. A loop will repeatedly update the value of `x` and keep printing it until a certain condition is met (or indefinitely if no condition is set).

## The while statement

Computers are often used to automate repetitive tasks. Repeating identical or similar tasks without making errors is something that computers do well and people do poorly. Because iteration is so common, Python provides several language features to make it easier.

One form of iteration in Python is the while statement. Here is a simple program that counts down from five and then says “Blastoff!”.

In [ ]:
n = 5
while n > 0:
    print(n)
    n = n - 1
print('Blastoff!')

5
4
3
2
1
Blastoff!


You can almost read the `while` statement as if it were English. It means, “While `n` is greater than 0, display the value of `n` and then reduce the value of `n` by 1. When you get to 0, exit the `while` statement and display the word `Blastoff`!”

More formally, here is the flow of execution for a `while` statement:

1. Evaluate the condition, yielding `True` or `False`.

2. If the condition is false, exit the `while` statement and continue execution at the next statement.

3. If the condition is true, execute the body and then go back to step 1.

This type of flow is called a *loop* because the third step loops back around to the top. We call each time we execute the body of the loop an *iteration*. For the above loop, we would say, “It had five iterations”, which means that the body of the loop was executed five times.

The body of the loop should change the value of one or more variables so that eventually the condition becomes false and the loop terminates. We call the variable that changes each time the loop executes and controls when the loop finishes the *iteration variable*. If there is no iteration variable, the loop will repeat forever, resulting in an *infinite loop*.

## Infinite loops
An endless source of amusement for programmers is the observation that the directions on shampoo, “Lather, rinse, repeat,” are an infinite loop because there is no iteration variable telling you how many times to execute the loop.

In the case of `countdown`, we can prove that the loop terminates because we know that the value of `n` is finite, and we can see that the value of `n` gets smaller each time through the loop, so eventually we have to get to 0. Other times a loop is obviously infinite because it has no iteration variable at all.

Sometimes you don’t know it’s time to end a loop until you get half way through the body. In that case you can write an infinite loop on purpose and then use the `break` statement to jump out of the loop.

This loop is obviously an *infinite* loop because the logical expression on the `while` statement is simply the logical constant `True`:

In [ ]:
# Warning: This loop is intentionally infinite. Use Stop/Interrupt to end it.
n = 10
while True:
    print(n, end=' ')
    n = n - 1
print('Done!')

This program will run forever or until your battery runs out because the logical expression at the top of the loop is always true by virtue of the fact that the expression is the constant value `True`.

While this is a dysfunctional infinite loop, we can still use this pattern to build useful loops as long as we carefully add code to the body of the loop to explicitly exit the loop using `break` when we have reached the exit condition.

For example, suppose you want to take input from the user until they type `done`. You could write:

In [ ]:
while True:
    line = input('> ')
    if line == 'done':
        break
    print(line)
print('Done!')

> >
>
> <
<
> 1
1
> a
a
> alewitj lknv
alewitj lknv
> done
Done!


The loop condition is `True`, which is always true, so the loop runs repeatedly until it hits the break statement.

Each time through, it prompts the user with an angle bracket. If the user types `done`, the `break` statement exits the loop. Otherwise the program echoes whatever the user types and goes back to the top of the loop.

This way of writing `while` loops is common because you can check the condition anywhere in the loop (not just at the top) and you can express the stop condition affirmatively (“stop when this happens”) rather than negatively (“keep going until that happens.”).

## Finishing iterations with continue

Sometimes you are in an iteration of a loop and want to finish the current iteration and immediately jump to the next iteration. In that case you can use the `continue` statement to skip to the next iteration without finishing the body of the loop for the current iteration.

Here is an example of a loop that copies its input until the user types “done”, but treats lines that start with the hash character as lines not to be printed (kind of like Python comments).

In [ ]:
while True:
    line = input('> ')
    if line[0] == '#':
        continue
    if line == 'done':
        break
    print(line)
print('Done!')

> a
a
> 123
123
> $10
$10
> # dont print this
> # abc 
> done
Done!


All the lines are printed except the one that starts with the hash sign because when the `continue` is executed, it ends the current iteration and jumps back to the `while` statement to start the next iteration, thus skipping the `print` statement.

## Looping Patterns
*Supplemental material*

### Counter controlled loops (while i < 10)

A counter-controlled loop is a loop that runs for a specified number of iterations, controlled by a counter variable that gets incremented or decremented each time the loop executes.

In [ ]:
# Simple Counter-Controlled Loop
i = 0  # Initialize the counter

while i < 10:  # Loop runs while i is less than 10
    print(f"Current value of i: {i}")
    i = i + 1  # Increment the counter by 1

Current value of i: 0
Current value of i: 1
Current value of i: 2
Current value of i: 3
Current value of i: 4
Current value of i: 5
Current value of i: 6
Current value of i: 7
Current value of i: 8
Current value of i: 9


In [ ]:
# Counter-Controlled Loop with a Step of 2
i = 0  # Initialize the counter

while i < 10:  # Loop runs while i is less than 10
    print(f"i is now: {i}")
    i = i + 2  # Increment the counter by 2


i is now: 0
i is now: 2
i is now: 4
i is now: 6
i is now: 8


In [ ]:
# Summing Numbers and Counting Evens
i = 1  # Initialize the counter
total_sum = 0  # Initialize the total sum
even_count = 0  # Initialize the even number counter

while i <= 10:  # Loop runs while i is less than or equal to 10
    total_sum += i  # Add the current value of i to total_sum
    if i % 2 == 0:  # Check if the current value of i is even
        even_count += 1  # Increment the even counter if i is even
    i += 1  # Increment the counter

# Print the results
print(f"Total sum of numbers from 1 to 10: {total_sum}")
print(f"Number of even numbers between 1 and 10: {even_count}")

Total sum of numbers from 1 to 10: 55
Number of even numbers between 1 and 10: 5


Explanation:
1. Counter Variable (`i`): Starts at `1` and increments by `1` each time through the loop.
2. Summing Numbers: Each iteration adds the current value of `i` to `total_sum`.
3. Counting Even Numbers: The `if` statement checks if i is even by using `i % 2 == 0`. If i is even, the `even_count` is incremented.
4. The loop stops when `i` exceeds `10`.

### Input validation using while True and break

Input validation using `while True` and `break` to repeatedly prompt the user until they provide valid input.

In [ ]:
# Validating a Number within a Range
while True:
    try:
        number = int(input("Enter a number between 1 and 100: "))
        if 1 <= number <= 100:  # Check if the number is within range
            print(f"Valid input: {number}")
            break  # Exit the loop when a valid number is entered
        else:
            print("Number out of range. Try again.")
    except ValueError:
        print("Invalid input. Please enter an integer.")

Enter a number between 1 and 100: abc
Invalid input. Please enter an integer.
Enter a number between 1 and 100: 200
Number out of range. Try again.
Enter a number between 1 and 100: -10
Number out of range. Try again.
Enter a number between 1 and 100: 50
Valid input: 50


Explanation:
1. `while True`: The loop runs indefinitely until a valid number is entered.
2. Input Validation: If the input is not a valid integer or not within the range 1-100, the program prints an error message and prompts the user again.
3. `break`: The loop exits once a valid number is provided.

In [ ]:
# Validating Yes/No Input
while True:
    response = input("Do you want to continue? (yes/no): ").lower()  # Convert input to lowercase
    if response == "yes" or response == "no":  # Check if the input is valid
        print(f"You entered: {response}")
        break  # Exit the loop when a valid response is entered
    else:
        print("Invalid input. Please enter 'yes' or 'no'.")

Do you want to continue? (yes/no): 10
Invalid input. Please enter 'yes' or 'no'.
Do you want to continue? (yes/no): python
Invalid input. Please enter 'yes' or 'no'.
Do you want to continue? (yes/no): abc
Invalid input. Please enter 'yes' or 'no'.
Do you want to continue? (yes/no): YES
You entered: yes


Explanation:
1. `while True`: The loop runs until a valid response ("yes" or "no") is provided.
2. Input Validation: The program checks if the input is either "yes" or "no". If not, it asks again.
3. `break`: The loop exits once the user enters a valid response.

### Summing, Counting, and Min/Max loops (with user input, NOT collections of numbers)

 Collections of numbers generally refer to data structures like lists, tuples, sets, or arrays that store multiple numbers. For example, a list like numbers = [5, 10, 15] is a collection of numbers. # We will cover this content in later weeks.



In [ ]:
# Simple Summing Loop with User Input
total = 0  # Initialize total to 0
count = 0  # Initialize counter

while count < 5:  # Repeat the loop 5 times
    try:
        number = float(input("Enter a number: "))  # Get user input
        total += number  # Add the number to the total
        count += 1  # Increment the counter
    except ValueError:
        print("Invalid input. Please enter a valid number.")

print(f"The total sum is: {total}")


Enter a number: 4.5
Enter a number: 3.2
Enter a number: -1
Enter a number: 2.5
Enter a number: 5
The total sum is: 14.2


Explanation:
- Summing: The `total += number` adds the inputted number to the `total` in each iteration.
- Counting: The `count += 1` increases the count after each valid input. This ensures the loop runs 5 times.
- Error Handling: If the user enters something invalid, an error message is displayed, but the loop continues asking for input.

Here's how the loop iterates through the process step-by-step, with each iteration updating the `total` and `count`:

### Initial State
- `total = 0`
- `count = 0`

### Iteration Process

**First Iteration:**
- `total = 0`
- `count = 0`
- User inputs a number (e.g., 4.5)
- `total = 0 + 4.5 = 4.5`
- `count = 0 + 1 = 1`

**Second Iteration:**
- `total = 4.5`
- `count = 1`
- User inputs another number (e.g., 3.2)
- `total = 4.5 + 3.2 = 7.7`
- `count = 1 + 1 = 2`

**Third Iteration:**
- `total = 7.7`
- `count = 2`
- User inputs another number (e.g., -1.0)
- `total = 7.7 + (-1.0) = 6.7`
- `count = 2 + 1 = 3`

**Fourth Iteration:**
- `total = 6.7`
- `count = 3`
- User inputs another number (e.g., 2.5)
- `total = 6.7 + 2.5 = 9.2`
- `count = 3 + 1 = 4`

**Fifth Iteration:**
- `total = 9.2`
- `count = 4`
- User inputs another number (e.g., 5.0)
- `total = 9.2 + 5.0 = 14.2`
- `count = 4 + 1 = 5`

### End of Loop
- Since `count` is now 5, which is equal to the loop condition, the loop exits.
- The final `total` is `14.2`.

The program prints:
```
The total sum is: 14.2
```

In [ ]:
# finding the minimum and maximum values using a loop
min_num = None
max_num = None

while True:
    user_input = input("Enter a number (or type 'done' to finish): ")

    if user_input == 'done':
        break  # Exit the loop if the user is done

    number = float(user_input)  # Convert input to a float

    # Update minimum and maximum values
    if min_num is None or number < min_num:
        min_num = number
    if max_num is None or number > max_num:
        max_num = number

# Check if any numbers were entered
if min_num is not None and max_num is not None:
    print(f"Minimum number: {min_num}")
    print(f"Maximum number: {max_num}")
else:
    print("No numbers were entered.")

Enter a number (or type 'done' to finish): 7.5
Enter a number (or type 'done' to finish): 3.2
Enter a number (or type 'done' to finish): 5.8
Enter a number (or type 'done' to finish): 4.1
Enter a number (or type 'done' to finish): done
Minimum number: 3.2
Maximum number: 7.5


Explanation:
1. Initialize `min_num` and `max_num` as `None` since no numbers have been entered yet.
2. Start a `while True` loop to repeatedly prompt the user to enter a number or type 'done' to stop.
3. If 'done' is entered, break the loop; otherwise, convert the input to a float.
4. Update `min_num` if it's the first number or smaller than the current value, and update `max_num` if it's the first or larger than the current value.
5. After the loop, check if any numbers were entered and print the min and max values, or notify if no numbers were entered.

Here is a breakdown of the loop's execution through 5 iterations:

Initial State:
* `min_num = None`
* `max_num = None`

Iteration Process:

1. First Iteration:
   * User input: 7.5
   * Convert to float: number = 7.5
   * min_num is None, so set min_num = 7.5
   * max_num is None, so set max_num = 7.5
   * State after iteration: min_num = 7.5, max_num = 7.5

2. Second Iteration:
   * User input: 3.2
   * Convert to float: number = 3.2
   * 3.2 < min_num (7.5), so set min_num = 3.2
   * 3.2 < max_num (7.5), so max_num remains 7.5
   * State after iteration: min_num = 3.2, max_num = 7.5

3. Third Iteration:
   * User input: 5.8
   * Convert to float: number = 5.8
   * 5.8 > min_num (3.2), so min_num remains 3.2
   * 5.8 < max_num (7.5), so max_num remains 7.5
   * State after iteration: min_num = 3.2, max_num = 7.5

4. Fourth Iteration:
   * User input: 4.1
   * Convert to float: number = 4.1
   * 4.1 > min_num (3.2), so min_num remains 3.2
   * 4.1 < max_num (7.5), so max_num remains 7.5
   * State after iteration: min_num = 3.2, max_num = 7.5

5. Fifth Iteration:
   * User input: done
   * This breaks the loop

Final State:
* min_num = 3.2
* max_num = 7.5

The program would then print:
```
Minimum number: 3.2
Maximum number: 7.5
```
